File: internal/cache/cache.go
package cache

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"

	"github.com/takara-ai/miru-code/internal/embeddings"
	"github.com/takara-ai/miru-code/internal/index"
	"github.com/takara-ai/miru-code/internal/utils"
)

func ResolveCacheFolder() string {
	name := "miru"
	home := os.Getenv("HOME")
	if home == "" {
		home = os.Getenv("USERPROFILE")
	}

	switch runtime.GOOS {
	case "windows":
		base := os.Getenv("LOCALAPPDATA")
		if base == "" {
			base = os.Getenv("APPDATA")
		}
		if base == "" {
			base = filepath.Join(home, "AppData", "Local")
		}
		return filepath.Join(base, name, "Cache")
	case "darwin":
		return filepath.Join(home, "Library", "Caches", name)
	default:
		if xdg := os.Getenv("XDG_CACHE_HOME"); xdg != "" {
			return filepath.Join(xdg, name)
		}
		return filepath.Join(home, ".cache", name)
	}
}

func FindIndexCachePath(path string) string {
	normalized, err := filepath.Abs(path)
	if err != nil {
		normalized = path
	}
	sum := sha256.Sum256([]byte(normalized))
	return filepath.Join(ResolveCacheFolder(), hex.EncodeToString(sum[:]), "index")
}

func ClearCache(path string) error {
	return os.RemoveAll(FindIndexCachePath(path))
}

func GetValidatedCache(path string, embeddingModel string, content []utils.ContentType) (string, bool) {
	indexPath := FindIndexCachePath(path)
	paths := index.PersistencePathsFor(indexPath)
	if missing := index.PathsExist(paths); len(missing) > 0 {
		return "", false
	}
	model := embeddingModel
	if model == "" {
		model = embeddings.ResolveEmbeddingModel()
	}
	raw, err := os.ReadFile(paths.Metadata)
	if err != nil {
		return "", false
	}
	var metadata map[string]any
	if err := json.Unmarshal(raw, &metadata); err != nil {
		return "", false
	}
	if !metadataMatches(metadata, model, content) {
		return "", false
	}
	semanticMetaRaw, err := os.ReadFile(filepath.Join(paths.SemanticIndex, "meta.json"))
	if err != nil {
		return "", false
	}
	var semanticMeta map[string]any
	if err := json.Unmarshal(semanticMetaRaw, &semanticMeta); err != nil {
		return "", false
	}
	if expectedDims, ok := embeddings.ResolveEmbeddingDimensions(model); ok && intFromAny(semanticMeta["dimensions"]) != expectedDims {
		return "", false
	}
	if !index.SemanticIndexMatchesStorage(paths.SemanticIndex) {
		return "", false
	}
	if rootPath := stringFromAny(metadata["root_path"]); rootPath != "" {
		if _, err := os.Stat(rootPath); err != nil {
			return "", false
		}
		for _, rel := range stringSliceFromAny(metadata["file_paths"]) {
			if _, err := os.Stat(filepath.Join(rootPath, rel)); err != nil {
				return "", false
			}
		}
	}
	return indexPath, true
}

func LoadCachedIndex(indexPath string) (index.IndexBundle, error) {
	return index.LoadIndexBundle(indexPath)
}

func metadataMatches(metadata map[string]any, embeddingModel string, content []utils.ContentType) bool {
	if stringFromAny(metadata["embedding_model"]) != embeddingModel {
		return false
	}
	if expectedDims, ok := embeddings.ResolveEmbeddingDimensions(embeddingModel); ok {
		storedDims := intFromAny(metadata["embedding_dimensions"])
		if storedDims != expectedDims {
			return false
		}
	}
	if index.SemanticStorageFromMetadata(metadata) != index.ResolveSemanticStorage() {
		return false
	}
	stored := stringSliceFromAny(metadata["content_type"])
	if len(stored) != len(content) {
		return false
	}
	set := map[string]bool{}
	for _, item := range stored {
		set[item] = true
	}
	for _, item := range content {
		if !set[string(item)] {
			return false
		}
	}
	return true
}

func stringFromAny(value any) string {
	if value == nil {
		return ""
	}
	if s, ok := value.(string); ok {
		return s
	}
	return ""
}

func intFromAny(value any) int {
	switch v := value.(type) {
	case float64:
		return int(v)
	case int:
		return v
	default:
		return 0
	}
}

func stringSliceFromAny(value any) []string {
	items, ok := value.([]any)
	if !ok {
		return nil
	}
	out := make([]string, 0, len(items))
	for _, item := range items {
		if s, ok := item.(string); ok {
			out = append(out, s)
		}
	}
	return out
}

File: src/cache.ts
import { createHash } from "node:crypto";
import { rm } from "node:fs/promises";
import { join } from "node:path";
import { resolveEmbeddingDimensions, resolveEmbeddingModel } from "./embeddings/openai.ts";
import {
  loadBm25,
  loadSemantic,
  pathsExist,
  persistencePaths,
  semanticIndexMatchesStorage,
} from "./index/persistence.ts";
import { resolveSemanticStorage, semanticStorageFromMetadata } from "./index/vector-storage.ts";
import type { Chunk, ContentType } from "./types.ts";
import { chunkFromDict } from "./types.ts";
import { computeSourceCacheKey } from "./utils.ts";
import { indexCacheEpoch } from "./version.ts";

export function resolveCacheFolder(): string {
  const override = process.env.MIRU_CACHE_HOME?.trim();
  if (override) {
    return override;
  }
  const name = "miru";
  const home = process.env.HOME ?? process.env.USERPROFILE ?? "";
  let base: string;
  if (process.platform === "win32") {
    base = process.env.LOCALAPPDATA ?? process.env.APPDATA ?? join(home, "AppData", "Local");
    base = join(base, name, "Cache");
  } else if (process.platform === "darwin") {
    base = join(home, "Library", "Caches", name);
  } else {
    base = process.env.XDG_CACHE_HOME
      ? join(process.env.XDG_CACHE_HOME, name)
      : join(home, ".cache", name);
  }
  return base;
}

export function findIndexCachePath(path: string, ref?: string | null): string {
  const normalized = computeSourceCacheKey(path, ref);
  const subdir = createHash("sha256").update(normalized, "utf-8").digest("hex");
  return join(resolveCacheFolder(), subdir, "index");
}

function metadataMatches(
  metadata: Record<string, unknown>,
  embeddingModel: string,
  content: ContentType[],
): boolean {
  try {
    if (metadata.index_epoch !== indexCacheEpoch()) {
      return false;
    }
    const stored = metadata.content_type as ContentType[];
    const model = metadata.embedding_model as string;
    if (model !== embeddingModel) {
      return false;
    }
    const expectedDims = resolveEmbeddingDimensions(embeddingModel);
    const storedDims = metadata.embedding_dimensions as number | undefined;
    if (expectedDims != null && storedDims !== expectedDims) {
      return false;
    }
    if (semanticStorageFromMetadata(metadata) !== resolveSemanticStorage()) {
      return false;
    }
    const a = new Set(stored);
    const b = new Set(content);
    if (a.size !== b.size) {
      return false;
    }
    for (const c of a) {
      if (!b.has(c)) {
        return false;
      }
    }
    return true;
  } catch {
    return false;
  }
}

export async function getValidatedCache(
  path: string,
  embeddingModel: string | undefined,
  content: ContentType[],
): Promise<string | null> {
  const indexPath = findIndexCachePath(path);
  const paths = persistencePaths(indexPath);
  const missing = await pathsExist(paths);
  if (missing.length > 0) {
    return null;
  }

  const model = embeddingModel ?? resolveEmbeddingModel();
  const metadata = JSON.parse(await Bun.file(paths.metadata).text()) as Record<string, unknown>;
  if (!metadataMatches(metadata, model, content)) {
    if (metadata.index_epoch !== indexCacheEpoch()) {
      await rm(indexPath, { recursive: true, force: true });
    }
    return null;
  }

  const semanticMeta = JSON.parse(await Bun.file(`${paths.semanticIndex}/meta.json`).text()) as {
    dimensions: number;
  };
  const expectedDims = resolveEmbeddingDimensions(model);
  if (expectedDims != null && semanticMeta.dimensions !== expectedDims) {
    return null;
  }
  if (!(await semanticIndexMatchesStorage(paths.semanticIndex))) {
    return null;
  }

  const rootPath = metadata.root_path as string | null;
  if (rootPath) {
    const filePaths = metadata.file_paths as string[];
    try {
      await Bun.file(rootPath).stat();
      for (const rel of filePaths) {
        try {
          await Bun.file(join(rootPath, rel)).stat();
        } catch {
          return null;
        }
      }
    } catch {
      return null;
    }
  }

  return indexPath;
}

export async function loadCachedIndex(indexPath: string): Promise<{
  bm25: Awaited<ReturnType<typeof loadBm25>>;
  semantic: Awaited<ReturnType<typeof loadSemantic>>;
  chunks: Chunk[];
  metadata: Record<string, unknown>;
}> {
  const paths = persistencePaths(indexPath);
  const bm25 = await loadBm25(paths.bm25Index);
  const semantic = await loadSemantic(paths.semanticIndex);
  const chunkData = JSON.parse(await Bun.file(paths.chunks).text()) as Record<string, unknown>[];
  const chunks = chunkData.map(chunkFromDict);
  const metadata = JSON.parse(await Bun.file(paths.metadata).text()) as Record<string, unknown>;
  return { bm25, semantic, chunks, metadata };
}

export async function clearCache(path: string): Promise<void> {
  const indexPath = findIndexCachePath(path);
  await rm(indexPath, { recursive: true, force: true });
}
